Mathlib/LinearAlgebra/BilinearForm/DualLattice.lean

```diff
@@ -0,0 +1,130 @@
+- `BilinForm.dualSubmodule_span_of_basis`: The dual of a lattice is spanned by the dual basis.
+
+## TODO
+Properly develop the material in the context of lattices.
+-/
+
+variable {R S M} [CommRing R] [Field S] [AddCommGroup M]
+variable [Algebra R S] [Module R M] [Module S M] [IsScalarTower R S M]
+
+namespace BilinForm
+
+variable (B : BilinForm S M)
+
+/-- The dual submodule of a submodule with respect to a bilinear form. -/
+def dualSubmodule (N : Submodule R M) : Submodule R M where
+  carrier := { x | ∀ y ∈ N, B x y ∈ (1 : Submodule R S) }
+  add_mem' {a b} ha hb y hy := by simpa using add_mem (ha y hy) (hb y hy)
+  zero_mem' y _ := by rw [B.zero_left]; exact zero_mem _
+  smul_mem' r a ha y hy := by
+    convert (1 : Submodule R S).smul_mem r (ha y hy)
+    rw [← IsScalarTower.algebraMap_smul S r a, bilin_smul_left, Algebra.smul_def]
+
+lemma mem_dualSubmodule {N : Submodule R M} {x} :
+    x ∈ B.dualSubmodule N ↔ ∀ y ∈ N, B x y ∈ (1 : Submodule R S) := Iff.rfl
+
+lemma le_flip_dualSubmodule {N₁ N₂ : Submodule R M} :
+    N₁ ≤ B.flip.dualSubmodule N₂ ↔ N₂ ≤ B.dualSubmodule N₁ := by
+  show (∀ (x : M), x ∈ N₁ → _) ↔ ∀ (x : M), x ∈ N₂ → _
+  simp only [mem_dualSubmodule, Submodule.mem_one, flip_apply]
+  exact forall₂_swap
+
+/-- The natural paring of `B.dualSubmodule N` and `N`.
+This is bundled as a bilinear map in `BilinForm.dualSubmoduleToDual`. -/
+noncomputable
+def dualSubmoduleParing {N : Submodule R M} (x : B.dualSubmodule N) (y : N) : R :=
+  (x.prop y y.prop).choose
+
+@[simp]
+lemma dualSubmoduleParing_spec {N : Submodule R M} (x : B.dualSubmodule N) (y : N) :
+    algebraMap R S (B.dualSubmoduleParing x y) = B x y :=
+  (x.prop y y.prop).choose_spec
+
+/-- The natural paring of `B.dualSubmodule N` and `N`. -/
+-- TODO: Show that this is perfect when `N` is a lattice.
+@[simps]
+noncomputable
+def dualSubmoduleToDual [NoZeroSMulDivisors R S] (N : Submodule R M) :
+    B.dualSubmodule N →ₗ[R] Module.Dual R N :=
+  { toFun := fun x ↦
+    { toFun := B.dualSubmoduleParing x
+      map_add' := fun x y ↦ NoZeroSMulDivisors.algebraMap_injective R S (by simp)
+      map_smul' := fun r m ↦ NoZeroSMulDivisors.algebraMap_injective R S
+        (by simp [← Algebra.smul_def]) }
+    map_add' := fun x y ↦ LinearMap.ext fun z ↦ NoZeroSMulDivisors.algebraMap_injective R S
+      (by simp)
+    map_smul' := fun r x ↦ LinearMap.ext fun y ↦ NoZeroSMulDivisors.algebraMap_injective R S
+      (by simp [← Algebra.smul_def]) }
+
+lemma dualSubmoduleToDual_injective (hB : B.Nondegenerate) [NoZeroSMulDivisors R S]
+    (N : Submodule R M) (hN : Submodule.span S (N : Set M) = ⊤) :
+    Function.Injective (B.dualSubmoduleToDual N) := by
+  intro x y e
+  ext
+  apply LinearMap.ker_eq_bot.mp hB.ker_eq_bot
+  apply LinearMap.ext_on hN
+  intro z hz
+  simpa using congr_arg (algebraMap R S) (LinearMap.congr_fun e ⟨z, hz⟩)
+
+lemma dualSubmodule_span_of_basis {ι} [Fintype ι] [DecidableEq ι] [FiniteDimensional S M]
```

Comment:
`FiniteDimensional S M` is automatic since you have a finite basis.

Reply:
But `dualBasis` needs `[FiniteDimensional S M]`. Or maybe we should remove that from `dualBasis`.

Reply:
You can write
```
let _ : FiniteDimensional S M := ...
```
before the `:` and it should be happy (not 100% sure the syntax is exactly that, but it is similar, and I forgot if we want `let` or `letI` here).

Reply:
I added that over at `dualBasis`.